src/city/resource.c

```diff
@@ -58,6 +61,37 @@ int city_resource_get_amount_including_granaries(resource_type resource, int amo
     return amount_stored;
 }
 
+int city_resource_get_available_empty_space_granaries(resource_type food, int respect_settings)
+{
+    int available_storage = 0;
+    if (!resource_is_food(food)) {
+        return 0;
+    }
+
+    for (building *b = building_first_of_type(BUILDING_GRANARY); b; b = b->next_of_type) {
+        if (!respect_settings) {
+            available_storage += b->resources[RESOURCE_NONE];
+        } else {
+            available_storage += building_granary_maximum_receptible_amount(food, b);
+        }
+    }
+
+    return available_storage;
+}
+
+int city_resource_get_available_empty_space_warehouses(resource_type resource, int respect_settings)
+{
+    int available_storage = 0;
+    for (building *b = building_first_of_type(BUILDING_WAREHOUSE); b; b = b->next_of_type) {
+        if (!respect_settings) {
+            available_storage += building_warehouse_get_amount(b, RESOURCE_NONE);
```

Comment:
Warehouses don't work like granaries. This will likely not work properly.

You need to find the total amount of resources stored on the granary (I think there's a function for that) and subtract `32` from that.

Reply:
During testing this worked (or at least I did not run across any problems during the tests.)

I based this code / decision on the code in ```int building_warehouse_accepts_storage(building *b, int resource, int *understaffed)```
Where it uses ```space->subtype.warehouse_resource_id == RESOURCE_NONE``` as part of the check to determine if a resource can be stored or not.

There are also other warehouse functions that set the resource_id to RESOURCE_NONE when a space is entirely empty.

But I realise now on looking at this again, it can double count empty spaces.
Let me rework this a bit.

Reply:
Updated.
Created a function in warehouse.c to count empty spaces or half filled spaces of the target resource.